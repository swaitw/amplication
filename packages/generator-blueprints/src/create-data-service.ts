import { DSGResourceData, FileMap } from "@amplication/code-gen-types";
import normalize from "normalize-path";
import { name as generatorName } from "../package.json";
import DsgContext from "./dsg-context";
import { EnumResourceType } from "./models";
import { prepareContext } from "./prepare-context";
import { createBlueprint } from "./blueprint/create-blueprint";
import { ILogger } from "@amplication/util-logging";
import { prepareDefaultPlugins } from "./utils/dynamic-installation/defaultPlugins";
import { dynamicPackagesInstallations } from "@amplication/dsg-utils";
import { AstNode } from "@amplication/csharp-ast";

export async function createDataService(
  dSGResourceData: DSGResourceData,
  internalLogger: ILogger,
  pluginInstallationPath?: string
): Promise<FileMap<AstNode>> {
  const context = DsgContext.getInstance;

  const { GIT_REF_NAME: gitRefName, GIT_SHA: gitSha } = process.env;
  await context.logger.info(
    `Running DSG ${generatorName} version: ${gitRefName} <${gitSha?.substring(
      0,
      7
    )}>`
  );

  dSGResourceData.pluginInstallations = prepareDefaultPlugins(
    dSGResourceData.pluginInstallations
  );

  await dynamicPackagesInstallations(
    dSGResourceData.pluginInstallations,
    pluginInstallationPath,
    internalLogger,
    context.logger
  );

  try {
    if (dSGResourceData.resourceType === EnumResourceType.MessageBroker) {
      internalLogger.info("No code to generate for a message broker");
      return null;
    }

    const startTime = Date.now();
    await prepareContext(
      dSGResourceData,
      internalLogger,
      pluginInstallationPath
    );

    await context.logger.info("Creating application...", {
      resourceId: dSGResourceData.resourceInfo.id,
      buildId: dSGResourceData.buildId,
    });

    const files = new FileMap<AstNode>(context.logger);

    await files.merge(await createBlueprint());

    // This code normalizes the path of each module to always use Unix path separator.
    await context.logger.info(
      "Normalizing modules path to use Unix path separator"
    );
    await files.replaceFilesPath((path) => normalize(path));

    const endTime = Date.now();
    internalLogger.info("Application creation time", {
      durationInMs: endTime - startTime,
    });

    await context.logger.info(
      "Creating application process finished successfully"
    );

    return files;
  } catch (error) {
    await internalLogger.error("Failed to run createDataService", {
      ...error,
    });
    throw error;
  }
}
