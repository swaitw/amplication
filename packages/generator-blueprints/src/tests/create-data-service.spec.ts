import { createDataService } from "../create-data-service";
import { TEST_DATA } from "./test-data";
import { getTemporaryPluginInstallationPath } from "./dynamic-plugin-installation-path";
import { rm } from "fs/promises";
import { ILogger } from "@amplication/util-logging";

jest.setTimeout(100000);

const temporaryPluginInstallationPath =
  getTemporaryPluginInstallationPath(__filename);

export const mockedLogger: ILogger = {
  debug: jest.fn(),
  info: jest.fn(),
  warn: jest.fn(),
  error: jest.fn(),
  child: jest.fn(() => mockedLogger),
};

describe("createBlueprint", () => {
  afterEach(async () => {
    jest.clearAllMocks();
    await rm(temporaryPluginInstallationPath, {
      recursive: true,
      force: true,
    });
  });

  test("generate blueprint code as expected", async () => {
    const files = await createDataService(
      TEST_DATA,
      mockedLogger,
      temporaryPluginInstallationPath
    );

    const pathToCode: {
      [k: string]: any;
    } = {};

    for await (const file of files.getAll()) {
      pathToCode[file.path] = file.code;
    }

    expect(pathToCode).toMatchSnapshot();
  });
});
