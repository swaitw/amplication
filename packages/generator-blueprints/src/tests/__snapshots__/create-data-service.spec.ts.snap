// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createBlueprint generate blueprint code as expected 1`] = `{}`;
